File: day24/src/main.py
from pathlib import Path


def parse_input(filename):
    section_inputs, section_gates = Path(filename).read_text().strip().split('\n\n')

    values = {
        input: int(value)
        for line in section_inputs.splitlines()
        for input, value in [line.split(': ')]
    }

    gates = {
        (op, in1, in2, out)
        for line in section_gates.splitlines()
        for inputs, out in [line.split(' -> ')]
        for op in ('AND', 'OR', 'XOR') if f' {op} ' in inputs
        for in1, in2 in [inputs.split(f' {op} ')]
    }

    return values, gates


def simulate(values, gates):
    remaining_gates = gates.copy()
    while ready_gates := {
        (op, in1, in2, out)
        for (op, in1, in2, out) in remaining_gates
        if in1 in values and in2 in values
    }:
        for gate in ready_gates:
            op, in1, in2, out = gate
            if op == 'AND':
                values[out] = values[in1] & values[in2]
            elif op == 'OR':
                values[out] = values[in1] | values[in2]
            elif op == 'XOR':
                values[out] = values[in1] ^ values[in2]

            remaining_gates.remove(gate)


def part1():
    values, gates = parse_input('my_input.txt')
    simulate(values, gates)
    binary_number = ''.join(str(values[k]) for k in reversed(sorted((k for k in values if k.startswith('z')))))
    print(f'Answer: {int(binary_number, 2)}')


def part2():
    _, gates = parse_input('my_input.txt')

    def out(s_op, s_in):
        for op, in1, in2, out in gates:
            if s_op == op and s_in in (in1, in2):
                return out

    def ins(s_out):
        for _, in1, in2, out in gates:
            if s_out == out:
                return {in1, in2}
        return {}

    # https://en.wikipedia.org/wiki/Adder_(electronics)#Full_adder
    # partN = xN XOR yN
    # fullN = xN AND yN
    # propN = partN AND carryN-1
    # carryN = fullN OR propN
    # sumN = partN XOR carryN-1 = zN

    max_bits = sum(1 for (*_, out) in gates if out.startswith('z'))
    swapped_outs = []
    carry_out = out('AND', 'x00')

    for bit_index in range(1, max_bits - 1):
        input, output = f'x{bit_index:02d}', f'z{bit_index:02d}'
        part_out = out('XOR', input)
        prop_out = out('AND', part_out) or out('AND', carry_out)
        sum_out = out('XOR', part_out) or out('XOR', carry_out)
        prev_sum_ins = ins(sum_out) | ins(prop_out)
        if carry_out not in prev_sum_ins:
            swapped_outs.append(carry_out)
        sum_out_ins = ins(sum_out)
        if part_out not in sum_out_ins:
            swapped_outs.append(part_out)
        expected_sum_out = {output}
        if sum_out not in expected_sum_out:
            swapped_outs.append(sum_out)
        full_out = out('AND', input)
        carry_out = out('OR', full_out) or out('OR', prop_out)
        carry_out_ins = ins(carry_out)
        if full_out not in carry_out_ins:
            swapped_outs.append(full_out)
        if prop_out not in carry_out_ins:
            swapped_outs.append(prop_out)

    print(f'Answer: {','.join(map(str, sorted(swapped_outs)))}')


if __name__ == '__main__':
    part1()
    part2()

File: day24/src/main.rs
use itertools::Itertools;
use std::collections::{HashMap, HashSet};
use std::fs;

type Values<'a> = HashMap<&'a str, u8>;
type Gates<'a> = HashSet<(&'a str, &'a str, &'a str, &'a str)>;

fn parse_input(input_text: &str) -> (Values, Gates) {
    let sections: Vec<&str> = input_text.trim().split("\n\n").collect();
    let section_inputs = sections[0];
    let section_gates = sections[1];

    let mut values = HashMap::new();
    for line in section_inputs.lines() {
        let parts: Vec<&str> = line.split(": ").collect();
        let key = parts[0];
        let value: u8 = parts[1].parse().unwrap();
        values.insert(key, value);
    }

    let mut gates = HashSet::new();
    for line in section_gates.lines() {
        let parts: Vec<&str> = line.split(" -> ").collect();
        let out = parts[1];
        for op in [" AND ", " OR ", " XOR "].iter() {
            if parts[0].contains(op) {
                let ins: Vec<&str> = parts[0].split(op).collect();
                let in1 = ins[0];
                let in2 = ins[1];
                gates.insert((op.trim(), in1, in2, out));
                break;
            }
        }
    }

    (values, gates)
}

fn simulate<'a>(values: &mut Values<'a>, gates: &Gates<'a>) {
    let mut remaining_gates = gates.clone();
    while !remaining_gates.is_empty() {
        let ready_gates: HashSet<_> = remaining_gates
            .iter()
            .filter(|&&(_, in1, in2, _)| values.contains_key(in1) && values.contains_key(in2))
            .cloned()
            .collect();

        for &(op, in1, in2, out) in &ready_gates {
            let val1 = values[in1];
            let val2 = values[in2];

            let result = match op {
                "AND" => val1 & val2,
                "OR" => val1 | val2,
                "XOR" => val1 ^ val2,
                _ => unreachable!(),
            };

            values.insert(gate.3, result);
            remaining_gates.remove(&gate);
            values.insert(out, result);
            remaining_gates.remove(&(op, in1, in2, out));
        }
    }
}

fn part1() {
    let input_text = fs::read_to_string("my_input.txt").unwrap();
    let (mut values, gates) = parse_input(&input_text);
    simulate(&mut values, &gates);

    let binary_number: String = values
        .iter()
        .filter(|(key, _)| key.starts_with('z'))
        .collect::<Vec<_>>()
        .into_iter()
        .sorted()
        .rev()
        .map(|(_, bit)| bit.to_string())
        .collect();

    let answer = usize::from_str_radix(&binary_number, 2).unwrap();
    println!("Answer: {}", answer);
}

fn part2() {
    let input_text = fs::read_to_string("my_input.txt").unwrap();
    let (_, gates) = parse_input(&input_text);

    let out = |s_op: &str, s_in: &str| -> Option<&str> {
        for &(op, in1, in2, out) in &gates {
            if op == s_op && (s_in == in1 || s_in == in2) {
                return Some(out);
            }
        }
        None
    };

    let ins = |s_out: &str| -> HashSet<&str> {
        for &(_, in1, in2, out) in &gates {
            if out == s_out {
                return HashSet::from([in1, in2]);
            }
        }
        HashSet::new()
    };

    let max_bits = gates
        .iter()
        .filter(|&&(_, _, _, out)| out.starts_with('z'))
        .count();

    // https://en.wikipedia.org/wiki/Adder_(electronics)#Full_adder
    // partN = xN XOR yN
    // fullN = xN AND yN
    // propN = partN AND carryN-1
    // carryN = fullN OR propN
    // sumN = partN XOR carryN-1 = zN

    let mut swapped_outs = Vec::new();
    let mut carry_out = out("AND", "x00").unwrap();

    for bit_index in 1..max_bits - 1 {
        let input = format!("x{:02}", bit_index);
        let part_out = out("XOR", &input).unwrap();
        let prop_out = out("AND", &part_out).or_else(|| out("AND", &carry_out)).unwrap();
        let sum_out = out("XOR", &part_out).or_else(|| out("XOR", &carry_out)).unwrap();
        let prev_sum_ins = &ins(&sum_out) | &ins(&prop_out);
        if !prev_sum_ins.contains(carry_out) {
            swapped_outs.push(carry_out);
        }
        let sum_out_ins = ins(&sum_out);
        if !sum_out_ins.contains(part_out) {
            swapped_outs.push(part_out);
        }
        let output = format!("z{:02}", bit_index);
        let expected_sum_out = HashSet::from([output.as_str()]);
        if !expected_sum_out.contains(sum_out) {
            swapped_outs.push(sum_out);
        }
        let full_out = out("AND", &input).unwrap();
        carry_out = out("OR", &full_out).or_else(|| out("OR", &prop_out)).unwrap();
        let carry_out_ins = ins(&carry_out);
        if !carry_out_ins.contains(full_out) {
            swapped_outs.push(full_out);
        }
        if !carry_out_ins.contains(prop_out) {
            swapped_outs.push(prop_out);
        }
    }

    println!("Answer: {}", swapped_outs.into_iter().sorted().join(","));
}

fn main() {
    part1();
    part2();
}
